### PR TITLE
Migrate kapt to ksp

### LINKED ISSUE #2470: demo application doesn't start on windows
**Describe the bug**
can't start the demo application on Windows 10 with java 17

Relevant log:

```
java.io.FileNotFoundException: C:\WINDOWS\sqlite-3.36.0-d1d580cc-94e1-44dd-976b-9d0203cd82d9-sqlitejdbc.dll.lck (Access is denied)
	at java.base/java.io.FileOutputStream.open0(Native Method)
	at java.base/java.io.FileOutputStream.open(FileOutputStream.java:298)
	at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:237)
	at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:187)
	at org.sqlite.SQLiteJDBCLoader.extractAndLoadLibraryFile(SQLiteJDBCLoader.java:210)
```

I managed to let it work by migrating to kspid("com.google.devtools.ksp").version("1.9.21-1.0.16"):
https://github.com/google/android-fhir/compare/master...icrc-fdeniger:android-fhir:icrc-ok-pull-data?expand=1#diff-84cb3f8fc96e6909e77e32e7f45a90496814fd34fb2e9822056aea55aaa0e9e7

I believe the root is that I can open a file located the folder  C:\WINDOWS ( not an admin)

### PR DESCRIPTION
**Pull request by icrc-fdeniger**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #2470

**Description**
To be polished.
- move stdlib to 1.9.22 ( to be compatible with ksp 1.9.22)
- use ksp instead of kapt

**Alternative(s) considered**
Have to run the demo in admin mode on windows but not possible due to our entreprise policy

**Type**
Bug fix

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by jingtang10** [comment]
thanks for this pr this looks fine to me if tests pass

**Review by jingtang10 — APPROVED** [review]

File: demo/build.gradle.kts [hunk 1]
```diff
@@ -1,7 +1,6 @@
 plugins {
   id(Plugins.BuildPlugins.application)
   id(Plugins.BuildPlugins.kotlinAndroid)
-  id(Plugins.BuildPlugins.kotlinKapt)
```
Comment (MJ1998) [inline demo/build.gradle.kts:4]: Is it needed to remove this or will it work even with this ?

File: engine/build.gradle.kts [hunk 2]
```diff
@@ -4,7 +4,7 @@ import java.net.URL
 plugins {
   id(Plugins.BuildPlugins.androidLib)
   id(Plugins.BuildPlugins.kotlinAndroid)
-  id(Plugins.BuildPlugins.kotlinKapt)
+  id(Plugins.BuildPlugins.kotlinKsp).version(Plugins.Versions.ksp)
```
Comment (MJ1998) [inline engine/build.gradle.kts:7]: Is it needed to apply the version here ? Can we not do something similar to the other Plugins being added here ?

File: knowledge/src/main/java/com/google/android/fhir/knowledge/db/dao/KnowledgeDao.kt [hunk 3]
```diff
@@ -63,7 +63,7 @@ abstract class KnowledgeDao {
   @Query(
     "DELETE from ResourceMetadataEntity WHERE resourceMetadataId NOT IN (SELECT DISTINCT resourceMetadataId from ImplementationGuideResourceMetadataEntity)",
   )
-  internal abstract suspend fun deleteOrphanedResources()
+  internal abstract suspend fun deleteOrphanedResources(): Int
```
Comment (MJ1998) [inline knowledge/src/main/java/com/google/android/fhir/knowledge/db/dao/KnowledgeDao.kt:66]: Is this causing an error with ksp? If not, then this is not needed to be changed in this PR.

**Review by MJ1998 — CHANGES_REQUESTED** [review]

File: demo/build.gradle.kts (hunk 1 above)
Comment (jingtang10) [reply demo/build.gradle.kts:4]: no we don't want unnecessary dependencies

File: knowledge/src/main/java/com/google/android/fhir/knowledge/db/dao/KnowledgeDao.kt (hunk 3 above)
Comment (icrc-fdeniger) [reply knowledge/src/main/java/com/google/android/fhir/knowledge/db/dao/KnowledgeDao.kt:66]: this modification has been done by @jingtang10. I suppose it's ok :)

**Comment by icrc-fdeniger** [comment]
push a new version with something similar to other plugins.
3 tests are failing with this error:
`java.lang.IllegalStateException: Illegal connection pointer 4411. Current pointers for thread Thread[arch_disk_io_1 @coroutine#204,5,SDK 24] []`
Will try to investigate today. don't know the root cause for now

**Comment by icrc-fdeniger** [comment]
I fixed some unit tests:

- NpmFileManagerTest: was failing on Windows due to File separator.
- FhirSyncWorkerTest: I add to start in test mode and clean up FhirEngineProvider to let the test pass. Without this modification I get the  error `java.lang.IllegalStateException: Illegal connection pointer 4411` 

Some tests are still failing and I suppose it's due to my locale:

- For instance QuestionnaireUiEspressoTest (expected: 2005-01-05T06:10 but was : 2005-01-05T18:10)

**Comment by MJ1998** [comment]
I believe the test failing in QuestionnaireUiEspressoTest is `dateTimePicker_shouldSetAnswerWhenDateAndTimeAreFilled`. It's a flaky test.  Even [this PR](https://github.com/google/android-fhir/pull/2326) has this problem.
@jingtang10 Is someone working on this flaky test ?

**Review by MJ1998 — APPROVED** [review]
LGTM. Thanks for testing these out on Windows.

**Comment by jingtang10** [comment]
> Some tests are still failing and I suppose it's due to my locale:

thanks for these! we shouldn't have tests that fail in different locales - we should fix those - can you point out which tests those are?

**Comment by jingtang10** [comment]
> I believe the test failing in QuestionnaireUiEspressoTest is `dateTimePicker_shouldSetAnswerWhenDateAndTimeAreFilled`. It's a flaky test. Even [this PR](https://github.com/google/android-fhir/pull/2326) has this problem. @jingtang10 Is someone working on this flaky test ?

got ya - can you take a look at this?

**Comment by vorburger** [comment]
FYI I have created #2485 re. the flaky `QuestionnaireUiEspressoTest$dateTimePicker_shouldSetAnswerWhenDateAndTimeAreFilled` test 12/24h problem.

**Comment by icrc-fdeniger** [comment]
Hello, thanks for your answers. Is there another step to let this PR be accepted ? I see that Kokoro step is still pending but I can't have the details.

**Comment by vorburger** [comment]
> Hello, thanks for your answers. Is there another step to let this PR be accepted ?

I (only now, didn't look before) noticed that it already has 2 approvals and auto-merge, so it just needs a green build, then it will go in.

> I see that Kokoro step is still pending but I can't have the details.

There were some problems which were just solved, see #2485 and #2488, I've tried to click _Update branch_ to see if it helps.

**Comment by vorburger** [comment]
@icrc-fdeniger following the merge of #2463 there is a small conflict - can you have a look and (manually) resolve that? Tx!

**Pull request merged**